### PR TITLE
test: unit test for project_ip_access_list resource

### PR DESCRIPTION
**Pull request by oarbusi**
## Description

unit test for project_ip_access_list resource

Link to any related issue(s): https://jira.mongodb.org/browse/CLOUDP-214982

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: internal/service/projectipaccesslist/resource_project_ip_access_list.go [hunk 1]
```diff
@@ -200,50 +200,15 @@ func (r *projectIPAccessListRS) Create(ctx context.Context, req resource.CreateR
...
-		ProjectID:        types.StringValue(projectIPAccessList.GetGroupId()),
-		CIDRBlock:        types.StringValue(projectIPAccessList.GetCidrBlock()),
-		IPAddress:        types.StringValue(projectIPAccessList.GetIpAddress()),
-		AWSSecurityGroup: types.StringValue(projectIPAccessList.GetAwsSecurityGroup()),
-		Comment:          types.StringValue(projectIPAccessList.GetComment()),
-		Timeouts:         projectIPAccessListModel.Timeouts,
-	}
-}
-
-func newMongoDBProjectIPAccessList(projectIPAccessListModel *tfProjectIPAccessListModel) *[]admin.NetworkPermissionEntry {
```
Comment (lantoli) [inline internal/service/projectipaccesslist/resource_project_ip_access_list.go:234]: is it ok to assume projectIPAccessListModel is never nil?

File: internal/service/projectipaccesslist/model_project_ip_access_list_test.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		"project_id": GroupID,
+	})
+)
+
+func TestNewMongoDBProjectIPAccessList(t *testing.T) {
+	testCases := []struct {
+		tfModel        *projectipaccesslist.TfProjectIPAccessListModel
+		expectedResult *[]admin.NetworkPermissionEntry
+		name           string
+	}{
```
Comment (lantoli) [inline internal/service/projectipaccesslist/model_project_ip_access_list_test.go:40]: i suppose it's ok to follow the same structure even if there is only one test

**Review by lantoli — APPROVED** [review]

File: internal/service/projectipaccesslist/resource_project_ip_access_list.go (hunk 1 above)
Comment (oarbusi) [reply internal/service/projectipaccesslist/resource_project_ip_access_list.go:234]: yes, it will never be null. Also, it's attributes `AWSSecurityGroup`,  `CIDRBlock` and `IPAddress` are also checked

File: internal/service/projectipaccesslist/model_project_ip_access_list_test.go (hunk 2 above)
Comment (oarbusi) [reply internal/service/projectipaccesslist/model_project_ip_access_list_test.go:40]: Kept it like this for consistency even though it's not needed

Comment (AgustinBettati) [reply internal/service/projectipaccesslist/model_project_ip_access_list_test.go:40]: think this is useful as it simplifies when a new test is added

File: internal/service/projectipaccesslist/model_project_ip_access_list_test.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+			sdkModel: &admin.NetworkPermissionEntry{
+				IpAddress: &IPAddress,
+				Comment:   &Comment,
+				GroupId:   &GroupID,
+			},
+			expectedResult: &projectipaccesslist.TfProjectIPAccessListModel{
+				ID:               types.StringValue(IPAddressID),
+				ProjectID:        types.StringValue(GroupID),
+				AWSSecurityGroup: types.StringValue(""),
+				CIDRBlock:        types.StringValue(""),
```
Comment (AgustinBettati) [inline internal/service/projectipaccesslist/model_project_ip_access_list_test.go:90]: Interesting that we are storing empty strings instead of defining terraform null value. Don't believe we need to take immediate action here, but nice to see how unit testing evidences this behaviour.

**Review by AgustinBettati — APPROVED** [review]
LGTM, small observation

File: internal/service/projectipaccesslist/model_project_ip_access_list_test.go (hunk 3 above)
Comment (oarbusi) [reply internal/service/projectipaccesslist/model_project_ip_access_list_test.go:90]: Yes its an interesting behaviour made explicit by the unit test. In this case, these attributes have `omitempty` (see [docs](https://pkg.go.dev/go.mongodb.org/atlas-sdk/v20231115002@v20231115002.1.0/admin#NetworkPermissionEntry)) so in this case when sent to the API they would be ignored (only one of `AWSSecurityGroup`, `CIDRBlock`, `IPAddress` can be set at the same time, as explained in the docs: `Your IP access list entry can be one **awsSecurityGroup**, one **cidrBlock**, or one **ipAddress**.` )

Comment (AgustinBettati) [reply internal/service/projectipaccesslist/model_project_ip_access_list_test.go:90]: actually given that these fields are pointers, the `omitempty` will work by including the property in the json if the pointer is not `nil` (even if it is a pointer to ""). In this case we are not including in the request because in https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/projectipaccesslist/resource_project_ip_access_list.go#L237 we define the field as nil if state has an empty string.
TLDR: it is working well, but if we use proper terraform null values the state would be more "correct", and simplify handling these fields in model conversions.
